### PR TITLE
interpolation functions

### PR DESCRIPTION
**Pull request by tthsqe12**
@whbart. This is of course not ready yet. I though the nmod dlogs would belong in `ulong_extras.h`, but the `nmod_t` is defined in `nmod_vec.h` which itself includes `ulong_extras.h`. Thus the nmod dlogs are in the logical place `nmod_vec/`. It might be worthwhile giving flint a massive cleanup soon.

I am also wondering where I should put the following types:
(1) `nmod_ctx_t` a proper context object for doing arithmetic modulo ulong's. This is currently `nmodf_ctx_t` in `nmod_mpoly.h` but it will get renamed and moved soon.
(2) `fmpz_mod_ctx_t` a proper context object for doing arithmetic modulo fmpz's. 
(3) `fmpz_mod_discrete_log_ph_t` the analogue of `nmod_discrete_log_ph_t` for arbitrary p.
(4) `nmod_bma_t` for running the berlekamp massey algorithm modulo a ulong.
(5) `fmpz_mod_bma_t` for running the berlekamp massey algorithm modulo a fmpz.

The functionality provided by `_fmpz_mod_poly_minpoly_bm` is not sufficient for the last two because it does not support lazy evaluation.

### CODE REVIEW

**Comment by tthsqe12** [comment]
@wbhart. Could you tell me what properties the matrix M has in the `_nmod_poly_hgcd`? There are no useful comments in the code and the documentation is equally useless. Specifically, I am wondering if it computes M, A, and B such that
(1) deg A >= (deg a)/2 > deg B
(2) M is a product of elementary matrices of the form [[Q 1] [1 0]]
The fact that it computes a sign as well as me nervous. If this is not what it computes, is it possible to get such an M, A, B from the output of `_nmod_poly_hgcd`?

**Comment by wbhart** [comment]
@tthsqe12 Unfortunately I do not recall. It does compute a transformation matrix, so something like 1 holds.

As for the sizes, I couldn't tell you. But I don't think 2 rings any bells.

Something like 3 should be right. There is a paper describing the algorithm of course (Thull-Yap). The algorithm has been modified slightly in Flint to save a matrix multiplication, if I recall correctly. Also, we use the same code for computing the resultant as well as the gcd, hence there is some extra bookkeeping code in there to compute the resultant.

**Comment by tthsqe12** [comment]
Without proper documentation, comments or test code on what is being computed, I am afraid I will have to write my own simple quadratic version. If you ever remember what that function is supposed to do, the proper function `nmod_poly_hgcd` can then be implemented in terms of that.

**Comment by wbhart** [comment]
Just in case you didn't realise, it's not computing the gcd. It's a half gcd function. A simple quadratic half gcd will have exactly the same properties if you compute the transformation matrix.

The reason (lena + 1)/2 is chosen is that this is the largest size any of those values will have. Since we swap the values about to save memory, all those values must have that size.

What the function actually computes and what is assumed about the inputs is documented in the Flint docs. How it computes the gcd and any proofs associated with that are in the Thull-Yap paper also mentioned in the docs. As I said, we made two minor adjustments, which should be relatively obvious when you look at the pseudocode in Thull-Yap. All the stuff relevant to resultants is flagged off with "if (res)" anyway, so that's not hard to isolate. As I said, we also move part of the algorithm to (possibly) save a matrix multiplication. This was not necessary in the original paper to get the complexity, but it is necessary in a practical implementation.

Unfortunately, this is just a fiendishly difficult function. I don't think the addition of any extra comments will clarify it. When I added the resultant stuff in 2014, I had to reconstruct what it did (bear in mind, I did not write the version that is there), and I was able to do that from the paper.

I vaguely recall the signs are just there to save having to change the signs of all the entries every time the matrix is updated, but don't quote me on that. That was the reason for the signs in the integer case, so it's probably why we did it here too.

**Comment by wbhart** [comment]
By the way, Thull-Yap is unpublished as far as I know. However, it is one of the few known-to-be-correct papers on the subject. We figured if they couldn't get their paper published, we certainly couldn't publish our trivial modifications.

**Comment by wbhart** [comment]
By pseudocode in Thull-Yap, I mean the pseudocode on page 5 of [1].

SgnR and SgnS in our code refer to the sign of the matrices which are called R and S in the paper. I think we also have an iterative case for performance reasons.

[1] http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.46.1941&rep=rep1&type=pdf

**Comment by wbhart** [comment]
And there is test code. It's t-hgcd. Perhaps you overlooked that. It tests precisely what the docs say the function does.

**Comment by tthsqe12** [comment]
Of course I know It is not calculating the gcd. If I wanted the gcd, then I would call gcd.
There is a thing called 'the half gcd' in the literature, which I have summarized in
https://github.com/tthsqe12/flint2/blob/discrete_log_ph/nmod_poly/bma.c#L17
The key is points (2) and (3) that I gave above. The flint documentation for _hgcd simply states that it produces a valid matrix identity. This on its own is useless unless (2) and (3) are satisfied. The _hgcd test code also faithfully simply tests that the matrix identity is true, but it has no tests on degrees. The comments in the test code seem to be wrong.

What is the sign of a matrix? Can the function here
https://github.com/tthsqe12/flint2/blob/discrete_log_ph/nmod_poly/bma.c#L34
be correctly and simply implemented in terms of _nmod_poly_hgcd?

**Comment by tthsqe12** [comment]
It turns out that the documentation for _nmod_poly_hcd and the test code comments were wrong, but the function itself seems to check out.  I have therefore assumed that it is a correct implementation of [1] and updated things accordingly.

**Comment by tthsqe12** [comment]
@wbhart Could you clarify the documentation on `fmpz_fdiv_qr_preinvn`?  The documentation says that it will be faster than itself when the number of limbs is greater than a symbol that I could not find in flint 😕. I am finding it 2x slower than the corresponding `fmpz_mod` (with no pre computation) at any limb size. I might be using it wrong.

**Comment by tthsqe12** [comment]
Also, it seems that barrett reduction mod p for 2 word p is about 3x faster than the simple a*b % p. I think this is a good result considering that I am seeing that the using the nmod reduction for a*b % p for single word p is only 1x - 2.3x faster than fmpz_mod.

**Comment by wbhart** [comment]
The documentation is obviously incorrect. You should compare with fmpz_fdiv_qr though.

I'm surprised to learn it is 2x slower now. Obviously wasn't the case when I wrote it. However, we don't use this everywhere because it was never much faster than straight division.

**Comment by tthsqe12** [comment]
@wbhart Could you let me know your thoughts on the interfaces?
The nmod interfaces are the end of the two docs
https://github.com/tthsqe12/flint2/blob/discrete_log_ph/doc/source/nmod_poly.rst
and
https://github.com/tthsqe12/flint2/blob/discrete_log_ph/doc/source/nmod_vec.rst
The code for berlekamp massey also has comments.

I know it will be desirable to have a version where you can specify the base of the logarithm and Fredrik will probably want one-use versions of berlekamp-massey/pohlig-hellman that construct one of these objects and then throws them away. These all can be accommodated.

I could also document the differences with the existing fmpz_mod_poly_minpoly.

**Comment by tthsqe12** [comment]
@wbhart assuming this passes I am satisfied with it.

**Comment by thofma** [comment]
At the moment I am using fmpz_preinv_t. Are you removing it? If so, what should I use instead?

**Comment by tthsqe12** [comment]
nono - I didn't retire fmpz_preinvn_t, I just removed it from the fmpz_mod_ctx_t since it is useless for the remainder operation. If you want the quotient and remainder or just the quotient, I have no idea how it compares.

**Comment by thofma** [comment]
OK.

**Comment by tthsqe12** [comment]
@wbhart Could you let me know what you need to merge this? I have a lot of code depending on this and have been waiting a long time to get this in flint.

**Comment by tthsqe12** [comment]
@thofma This last commit has the documentation for fmpz_mod. I have decided to make a divides function like most of the other rings in flint. If you don't like the interface, let me know.

**Comment by thofma** [comment]
Best interface ever. I hope this gets merged soon!

**Comment by tthsqe12** [comment]
Well, now I have to make some implementation changes, but it should be done soon.

**Comment by fredrik-johansson** [comment]
My only comment is that the fmpz uiui/uiuiui functions are too heavyweight to be inlines.

File: doc/source/fmpz_mod.rst [hunk 1]
```diff
@@ -0,0 +1,104 @@
+.. _fmpz-mod:
+
+**fmpz_mod.h** -- arithemtic modulo integers
```
Comment (fredrik-johansson) [inline doc/source/fmpz_mod.rst]: spleling

**Comment by tthsqe12** [comment]
@fredrik-johansson the _fmpz_mod_mul2 function operates on the nanosecond scale and is noticeably impacted by the function call overhead. I would therefore like, with permission, to keep get_uiui and set_uiui inline like neg_uiui is. I have moved the uiuiui calls to ui_array calls though.

**Comment by tthsqe12** [comment]
I really can't find anything else to polish here.

**Comment by tthsqe12** [comment]
@wbhart I have a couple of branches that have been backed up for some time now because of these functions. I would appreciate if you could look at this PR soon.

**Comment by wbhart** [comment]
I can merge this, but I have one comment. Could the flint_aborts at least have the memory cleaned up before they are called (just the memory allocated in these functions of course, not in the caller).

**Comment by tthsqe12** [comment]
Can you give a specific location?

**Comment by wbhart** [comment]
You can just do a search for flint_abort and flint_throw in the "Files changed" page on the PR:

https://github.com/wbhart/flint2/pull/553

**Comment by tthsqe12** [comment]
I will replace all of the abort functions with the the flint_throw since that is what I meant to use.
While I certainly can clean up local variables, I am not completely sure what you mean by cleanup. When a function is modifying objects A,B,... and it throws, is the intention to have A, B, ... in valid states before the throw function is called?

**Comment by wbhart** [comment]
No, I just mean if there is an fmpz_init, say, there should be a corresponding fmpz_clear before the flint_throw, i.e. just clean up as much allocated memory as is possible.

**Comment by tthsqe12** [comment]
Ok. But, the memory allocated in these functions is probably a drop in the bucket compared to what was allocated down the stack.

**Comment by wbhart** [comment]
Sure. But especially with out memory manager, it pays to clean up as much as we can.

**Comment by tthsqe12** [comment]
Thank you so much.

**Pull request merged**